src/shared/metrics.c

```diff
@@ -0,0 +1,186 @@
+        sd_varlink_reply_flags_t flags,
+        void *userdata) {
+
+        int *ret = ASSERT_PTR(userdata), r;
+
+        assert(link);
+
+        if (error) {
+                /* If we can translate this to an errno, let's print that as errno and return it, otherwise, return a generic error code */
+                r = sd_varlink_error_to_errno(error, parameters);
+                if (r != -EBADR)
+                        *ret = log_error_errno(r, "Method call failed: %m");
+                else
+                        r = *ret = log_error_errno(SYNTHETIC_ERRNO(EBADE), "Method call failed: %s", error);
+        } else
+                r = 0;
+
+        sd_json_variant_dump(parameters, SD_JSON_FORMAT_PRETTY_AUTO|SD_JSON_FORMAT_COLOR_AUTO, stdout, NULL);
+
+        return r;
+}
+
+static int metrics_call(const char *path) {
+        _cleanup_(sd_varlink_unrefp) sd_varlink *vl = NULL;
+        int r;
+
+        assert(path);
+
+        r = sd_varlink_connect_address(&vl, path);
+        if (r < 0)
+                return log_debug_errno(r, "Unable to connect to %s: %m", path);
+
+        int ret = 0;
+        sd_varlink_set_userdata(vl, &ret);
+
+        r = sd_varlink_bind_reply(vl, metrics_on_query_reply);
+        if (r < 0)
+                return log_debug_errno(r, "Failed to bind reply callback: %m");
+
+        sd_varlink_observe(vl, "io.systemd.Metrics.List", /* parameter */ NULL);
+        if (r < 0)
+                return log_debug_errno(r, "Failed to invoke varlink method: %m");
+
+        for (;;) {
+                r = sd_varlink_is_idle(vl);
+
+                if (r < 0)
+                        return log_error_errno(r, "Failed to check if varlink connection is idle: %m");
+                if (r > 0) {
+                        break;
+                }
+
+                r = sd_varlink_process(vl);
+                if (r < 0)
+                        return log_error_errno(r, "Failed to process varlink connection: %m");
+                if (r != 0)
+                        continue;
+
+                r = sd_varlink_wait(vl, USEC_INFINITY);
+                if (r < 0)
+                        return log_error_errno(r, "Failed to wait for varlink connection events: %m");
+        }
+
+        return ret;
+}
+
+static int metrics_start_query(void) {
+        _cleanup_closedir_ DIR *d = NULL;
+        int r;
+
+        d = opendir("/run/systemd/metrics/");
+        if (!d) {
+                if (errno == ENOENT)
+                        return -ESRCH;
+
+                return -errno;
+        }
+
+        FOREACH_DIRENT(de, d, return -errno) {
+                _cleanup_free_ char *p = NULL;
+
+                p = path_join("/run/systemd/metrics/", de->d_name);
+                if (!p)
+                        return -ENOMEM;
+
+                r = metrics_call(p);
+                if (r < 0)
+                        return r;
+        }
+
+        return 0;
+}
+
+int metrics_query_all(sd_json_variant **ret) {
```

Comment:
this function and all relevant should not be in shared. Since this code will only be used in CLI it doesn't make sense the share it. Please make it in repart.c (or somewhere there, not in shared)